### PR TITLE
How libstdc++ `std::unordered_map` implemented?

### PR DESCRIPTION
**Pull request by ilvokhin**

### CODE REVIEW

File: posts/libstdc++-std-unordered-map/libstdc++-unordered-map.md [hunk 1]
```diff
@@ -0,0 +1,613 @@
+We all love maps. We love hash maps even more. They should be fast and
+help to solve a large number of problems. Do you ever wonder about how they are
+working under the hood? In this post I am going to explore implementation
+details of unordered associative containers from C++ Standard Library
+(precisely GCC's libstdc++ implementation).
```
Comment (semenovaSN) [inline posts/libstdc++-std-unordered-map/libstdc++-unordered-map.md]: Maybe worth to add a specific version?

File: posts/libstdc++-std-unordered-map/libstdc++-unordered-map.md [hunk 2]
```diff
@@ -0,0 +1,613 @@
...
+The `_Hashtable` class itself is a combination of
+`std::forward_list<_Hash_node>` containing the elements and
+`std::vector<std::forward_list<_Hash_node>::iterator>` representing the buckets
+([code comment][11]).
+
+`_Hash_node_base** _M_buckets` is an array of pointers to hash table nodes. You
+can think of it as `_Hash_node_base* _M_buckets[]` instead of pointer to a
+pointer.
+
+`_Hash_node_base _M_before_begin` is a special sentinel node without any user
```
Comment (semenovaSN) [inline posts/libstdc++-std-unordered-map/libstdc++-unordered-map.md]: Can we add a link to a sentinel pattern description?

File: posts/libstdc++-std-unordered-map/libstdc++-unordered-map.md [hunk 3]
```diff
@@ -0,0 +1,613 @@
...
+  const __rehash_state& __saved_state = _M_rehash_policy._M_state();
+  std::pair<bool, std::size_t> __do_rehash
+= _M_rehash_policy._M_need_rehash(_M_bucket_count, _M_element_count,
+                  __n_elt);
+
+  if (__do_rehash.first)
+{ 
+  _M_rehash(__do_rehash.second, __saved_state);
+  __bkt = _M_bucket_index(__code);
+}
```
Comment (semenovaSN) [inline posts/libstdc++-std-unordered-map/libstdc++-unordered-map.md]: Should the brackets be shifted here?

File: posts/libstdc++-std-unordered-map/libstdc++-unordered-map.md [hunk 4]
```diff
@@ -0,0 +1,613 @@
...
+```cpp
+  __node_ptr
+  _M_find_node(size_type __bkt, const key_type& __key,
+       __hash_code __c) const
+  {    
+__node_base_ptr __before_n = _M_find_before_node(__bkt, __key, __c);
+if (__before_n)
+  return static_cast<__node_ptr>(__before_n->_M_nxt);
+return nullptr;
+  }
```
Comment (semenovaSN) [inline posts/libstdc++-std-unordered-map/libstdc++-unordered-map.md]: Should the code be shifted here as well?

File: posts/libstdc++-std-unordered-map/libstdc++-unordered-map.md [hunk 5]
```diff
@@ -0,0 +1,613 @@
...
+`_M_find_before_node` does mostly what we expect it to do, but has a couple of
+interesting things in the sleeve.
+
+We [locate][23] a pointer to the element before the first bucket element.
+
+```cpp
+  __node_base_ptr __prev_p = _M_buckets[__bkt];
+  if (!__prev_p)
+return nullptr;
+```
```
Comment (semenovaSN) [inline posts/libstdc++-std-unordered-map/libstdc++-unordered-map.md]: Is this code correct? Feels like return under if [is missed ](https://github.com/gcc-mirror/gcc/blob/b9b7981f3d6919518372daf4c7e8c40dfc58f49d/libstdc%2B%2B-v3/include/bits/hashtable.h#L817)and should we shift the `return nullptr;`?

File: posts/libstdc++-std-unordered-map/libstdc++-unordered-map.md [hunk 6]
```diff
@@ -0,0 +1,613 @@
...
+  for (__node_ptr __p = static_cast<__node_ptr>(__prev_p->_M_nxt);;
+   __p = __p->_M_next())
+{ 
+  if (this->_M_equals(__k, __code, *__p))
+    return __prev_p;
+  
+  if (!__p->_M_nxt || _M_bucket_index(*__p->_M_next()) != __bkt)
+    break;
+  __prev_p = __p;
+}
```
Comment (semenovaSN) [inline posts/libstdc++-std-unordered-map/libstdc++-unordered-map.md]: Should brackets be shifted?

File: posts/libstdc++-std-unordered-map/libstdc++-unordered-map.md [hunk 7]
```diff
@@ -0,0 +1,613 @@
...
+  if (size() <= __small_size_threshold())
+{
+  __prev_n = _M_find_before_node(__k);
+  if (!__prev_n)
+    return 0;
+
+  // We found a matching node, erase it.
+  __n = static_cast<__node_ptr>(__prev_n->_M_nxt);
+  __bkt = _M_bucket_index(*__n);
+}
```
Comment (semenovaSN) [inline posts/libstdc++-std-unordered-map/libstdc++-unordered-map.md]: Brackets shift

File: posts/libstdc++-std-unordered-map/libstdc++-unordered-map.md [hunk 8]
```diff
@@ -0,0 +1,613 @@
...
+  __bkt = _M_bucket_index(__code);
+
+  // Look for the node before the first matching node.
+  __prev_n = _M_find_before_node(__bkt, __k, __code);
+  if (!__prev_n)
+    return 0;
+
+  // We found a matching node, erase it.
+  __n = static_cast<__node_ptr>(__prev_n->_M_nxt);
+}
```
Comment (semenovaSN) [inline posts/libstdc++-std-unordered-map/libstdc++-unordered-map.md]: Brackets shift

File: posts/libstdc++-std-unordered-map/libstdc++-unordered-map.md [hunk 9]
```diff
@@ -0,0 +1,613 @@
...
+  if (__prev_n == _M_buckets[__bkt])
+_M_remove_bucket_begin(__bkt, __n->_M_next(),
+  __n->_M_nxt ? _M_bucket_index(*__n->_M_next()) : 0);
+  else if (__n->_M_nxt)
+{
+  size_type __next_bkt = _M_bucket_index(*__n->_M_next());
+  if (__next_bkt != __bkt)
+    _M_buckets[__next_bkt] = __prev_n;
+}
+
```
Comment (semenovaSN) [inline posts/libstdc++-std-unordered-map/libstdc++-unordered-map.md:551]: Shift?

File: posts/libstdc++-std-unordered-map/libstdc++-unordered-map.md [hunk 10]
```diff
@@ -0,0 +1,613 @@
...
+## Closing thoughts
+
+There are a lot of cool tricks implemented to improve performance of
+`std::unordered_map` libstdc++ implementation. These tricks help for sure,
+but the main issue of `std::unordered_map` (not only libstdc++ implementation,
+but all standard compatible ones) is cache unfriendliness.
+
+Almost every operation on the container is practically a textbook example of
+[pointer chasing][32], so for real word use cases performance would not be as great
```
Comment (semenovaSN) [inline posts/libstdc++-std-unordered-map/libstdc++-unordered-map.md]: Did you mean `real world`?

File: posts/libstdc++-std-unordered-map/libstdc++-unordered-map.md (hunk 2 above)
Comment (ilvokhin) [reply posts/libstdc++-std-unordered-map/libstdc++-unordered-map.md]: Removed sentinel from text at all, not sure it was used correctly here.

File: posts/libstdc++-std-unordered-map/libstdc++-unordered-map.md (hunk 5 above)
Comment (ilvokhin) [reply posts/libstdc++-std-unordered-map/libstdc++-unordered-map.md]: Yeah, code is correct, but the reference link is wrong. Should be this one: https://github.com/gcc-mirror/gcc/blob/b9b7981f3d6919518372daf4c7e8c40dfc58f49d/libstdc%2B%2B-v3/include/bits/hashtable.h#L1939-L1941

File: posts/libstdc++-std-unordered-map/libstdc++-unordered-map.md (hunk 10 above)
Comment (ilvokhin) [reply posts/libstdc++-std-unordered-map/libstdc++-unordered-map.md]: Yes :)

**Review by ilvokhin — COMMENTED** [review]
Thanks for review, all comments were addressed.

**Pull request closed without merging**